# `nf-core/demo`

[`nf-core/demo`](https://nf-co.re/demo/) is a simple nf-core style bioinformatics pipeline for workshops and demonstrations.

It was created using the nf-core template and is designed to run quickly and demonstrate customization options.

<figure class="excalidraw">
--8<-- "docs/nf_customize/img/subway.excalidraw.svg"
</figure>

The [`nf-core/demo`](https://nf-co.re/demo/) pipeline consists of three processes:

1. ([`FASTQC`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)): Read QC
2. ([`FASTP`](https://github.com/OpenGene/fastp)): Adapter and quality trimming
3. ([`MULTIQC`](http://multiqc.info/)): Present QC for raw reads

It will produce four output folders:

1. `fastqc/`
    - `*_fastqc.html`: FastQC report containing quality metrics.
    - `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.
2. `fastp/`
    - `*.fastp.html`: Trimming report in html format.
    - `*.fastp.json`: Trimming report in json format.
    - `*.fastp.log`: Trimming log file.
    - `*.fastq.gz`: If --save_trimmed
3. `multiqc/`
    - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
    - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
    - `multiqc_plots/`: directory containing static images from the report in various formats.
4. `pipeline_info/`
    - Reports generated by Nextflow
    - Reports generated by nf-core
    - Parameters file

Each nf-core pipeline comes with extensive usage documentation about the pipeline, its parameters, its usage, and its outputs.

You can explore the documentation for the `nf-core/demo` pipeline [here](https://nf-co.re/demo/docs/usage).

## Your first `nf-core/demo` execution command

Each nf-core pipeline comes with extensive usage documentation to help you understand the requirements for running each pipeline.

You can explore the usage documentation for the `nf-core/demo` pipeline [here](https://nf-co.re/demo/docs/usage).

Notably, the `nf-core/demo` pipeline **requires** two parameters (`--input` and `--outdir`) to run.

Without these, the pipeline will not launch and nextflow will throw an error.

### `--input`

The `--input` parameter requires a path to comma-separated file containing information about the samples in the experiment.

```bash
--input 'path/to/samplesheet.csv'
```

Using the nf-core/demo usage documentation, you can see that it requires a comma-separated file (`.csv`) with 3 columns and a header row. The samplesheet file may consist of both single- and paired-end data may look something like the one below.

This is for 3 samples:

```csv title="samplesheet.csv"
sample,fastq_1,fastq_2
SAMPLE1_PE,path/to/sample1_R1.fastq.gz,path/to/sample1_R2.fastq.gz
SAMPLE2_PE,path/to/sample2_R1.fastq.gz,path/to/sample2_R2.fastq.gz
SAMPLE3_SE,path/to/sample3_R1.fastq.gz,
```

The pipeline will auto-detect whether a sample is single- or paired-end and if a sample has been sequenced more than once using the information provided in the samplesheet.

!!! question "Exercise"

    Create a samplesheet named `samplesheet.csv` for the pair end reads for samples gut, liver, and lung in the `data` folder.

    First, create the `samplesheet.csv`:

    ```bash
    code samplesheet.csv
    ```

    Next, fill is with the header line, and for each sample, an id and the complete paths to the forward and reverse reads:

    ```csv title="samplesheet.csv"
    sample,fastq_1,fastq_2
    gut,/workspace/gitpod/nf-customize/data/gut_1.fq.gz,/workspace/gitpod/nf-customize/data/gut_2.fq.gz
    liver,/workspace/gitpod/nf-customize/data/liver_1.fq.gz,/workspace/gitpod/nf-customize/data/liver_2.fq.gz
    lung,/workspace/gitpod/nf-customize/data/lung_1.fq.gz,/workspace/gitpod/nf-customize/data/lung_2.fq.gz
    ```

    Make sure it is saved in your working directory (`/workspace/gitpod/nf-customize/`)

### `--outdir`

The `--output` parameter is used to name the output directory where the results will be saved.

```bash
--output results
```

!!! question "Exercise"

    Execute the `nf-core/demo` pipeline with your new samplesheet as an input and the output directory named `results`"

    If you have saved your samplesheet in your working directory then you do not need to supply a full path.

    ```bash
    nextflow run nf-core/demo -r dev --input samplesheet.csv --outdir results
    ```

The above exercise is expected to fail as the software required to run each process is not available.

```console
ERROR ~ Error executing process > 'NFCORE_DEMO:DEMO:FASTP (sample1)'

Caused by:
  Process `NFCORE_DEMO:DEMO:FASTP (sample1)` terminated with an error exit status (127)
<truncated>
```

Fortunately, nf-core pipelines come packed with directives for containers and environments that can be flexibly enabled.

### Profiles

```bash
-profile <profile name>
```

Here, we will can add the `singularity` profile to our execution command, which will download software images to run each process for us.

!!! Exercise

    Execute the command again, but this time with the singularity profile:

    ```bash
    nextflow run nf-core/demo -r dev --input samplesheet.csv --outdir results -profile singularity
    ```

    The `nf-core/demo` pipeline should now run successfully.
